### PR TITLE
Daniandrew/APPEALS-67327

### PR DESCRIPTION
**Pull request by drew-hadley**
Resolves [Caseflow: Update Remediation job to use Status column for processing backlog](https://jira.devops.va.gov/browse/APPEALS-67327)

# Description

- Added a migration `remediation_status` column to the `EventRecords` table that will use this enum
```ruby
enum remediation_status: {
    pending: 0, # the event_record has not yet been processed by the job
    processed: 1, # the event_record was processed by the job without need for remediation
    remediated: 2, # the event_record was processed by the job and successfully remediated
    failed: 3 # the event_record was processed by the job, but the remediation failed
  }
```
- Added a migration `remediation_attempts` column to the `EventRecords` table to keep a count of how many times an event_record is tried. 
- Updated the following classes to utilize the new `EventRecord` attributes:
  -  `app/models/event_record.rb`
  - `app/jobs/person_and_veteran_event_remediation_job.rb`
  - `app/services/remediations/duplicate_person_remediation_service.rb`
  - `app/services/remediations/veteran_record_remediation_service.rb`
- Added tests for all of the above. 

Also includes some refactors to address code climate errors that came up [here](https://codeclimate.com/github/department-of-veterans-affairs/caseflow/pull/23775#)

## Acceptance Criteria
- [x] Code compiles correctly

# Backend
## Database Changes
*Only for Schema Changes*

* [ ] Add typical timestamps (`created_at`, `updated_at`) for new tables
* [ ] Update column comments; include a "PII" prefix to indicate definite or potential [PII data content](https://github.com/department-of-veterans-affairs/appeals-team/blob/main/caseflow-team/0-how-we-work/pii-handbook.md#what-is-pii)
* [x] Have your migration classes inherit from `Caseflow::Migration`, especially when adding indexes (use `add_safe_index`) (see [Writing DB migrations](https://github.com/department-of-veterans-affairs/caseflow/wiki/Writing-DB-migrations))
* [ ] Verify that `migrate:rollback` works as desired ([`change` supported functions](https://edgeguides.rubyonrails.org/active_record_migrations.html#using-the-change-method))
* [ ] Perform query profiling (eyeball Rails log, check bullet and fasterer output)
* [ ] For queries using raw sql was an explain plan run by System Team
* [ ] Add appropriate indexes (especially for foreign keys, polymorphic columns, unique constraints, and Rails scopes)
* [ ] Run `make check-fks`; add any missing foreign keys or add to `config/initializers/immigrant.rb` (see [Record associations and Foreign Keys](https://github.com/department-of-veterans-affairs/caseflow/wiki/Record-associations-and-Foreign-Keys))
* [ ] Add `belongs_to` for associations to enable the [schema diagrams](https://department-of-veterans-affairs.github.io/caseflow/task_trees/schema/schema_diagrams) to be automatically updated
* [ ] Document any non-obvious semantics or logic useful for interpreting database data at [Caseflow Data Model and Dictionary](https://github.com/department-of-veterans-affairs/caseflow/wiki/Caseflow-Data-Model-and-Dictionary)

# Best practices
## Code Documentation Updates
- [x] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [x] RSpec
- [ ] Jest
- [ ] Other

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [x] No new code climate issues added

### CODE REVIEW

File: app/services/remediations/duplicate_person_remediation_service.rb [hunk 1]
```diff
@@ -39,31 +46,32 @@ def call
...
-          after_data: record.attributes,
-          before_data: before_data,
-          id: record.id,
-          type: record.class.name
+        # Use the parameter object for the audit params
+        audit_params = RemediationAuditParams.new(
+          record.attributes,  # after_data
+          before_data,        # before_data
+          record.class.name,  # type
+          record.id           # id
```
Comment (raymond-hughes) [inline app/services/remediations/duplicate_person_remediation_service.rb:55]: No reason to keep all of the comments. Lets get it all cleaned up

**Review by raymond-hughes — CHANGES_REQUESTED** [review]

File: app/models/events/event_record.rb [hunk 2]
```diff
@@ -6,6 +6,18 @@ class EventRecord < CaseflowRecord
...
   store_accessor :info
 
+  enum remediation_status: {
+    pending: 0, # the event_record has not yet been processed by the job
+    processed: 1, # the event_record was processed by the job without need for remediation
+    remediated: 2, # the event_record was processed by the job and successfully remediated
+    failed: 3 # the event_record was processed by the job, but the remediation failed
+  }
+
+  # scope :pending, -> { where(remediation_status: pending) }
```
Comment (ryanpmessner) [inline app/models/events/event_record.rb:16]: Were you just testing these and left them in?  Should probably remove if not used.

File: app/services/remediations/veteran_record_remediation_service.rb [hunk 3]
```diff
@@ -3,22 +3,43 @@
...
     @event_record = event_record
   end
 
   def remediate!
-    if dups.any?
-      # If there are duplicates, run dup_fix on @after_fn
+    if dups.any? # If there are duplicates, run dup_fix on @after_fn
       if dup_fix(after_fn)
         dups.each(&:destroy!)
+        @event_record.remediated!
```
Comment (ryanpmessner) [inline app/services/remediations/veteran_record_remediation_service.rb:35]: there's an `attr_reader` for this variable, you might end up w/ an 'assumes too much' codeclimate error for it if you use the @variable directly

**Review by ryanpmessner — CHANGES_REQUESTED** [review]
Just the one change w/ the `@event_record` variable

File: app/services/remediations/duplicate_person_remediation_service.rb [hunk 4]
```diff
@@ -18,7 +20,11 @@ def initialize(updated_person_id:, duplicate_person_ids:, event_record:)
   def remediate!
     if find_and_update_records
       duplicate_persons.each(&:destroy!)
+      @event_record.remediated!
```
Comment (ryanpmessner) [inline app/services/remediations/duplicate_person_remediation_service.rb:23]: sorry, same deal here as w/ the veteran service, must not have noticed it before

File: app/services/remediations/veteran_record_remediation_service.rb (hunk 3 above)
Comment (dcoleman21) [reply app/services/remediations/veteran_record_remediation_service.rb]: Changed

**Review by ryanpmessner — APPROVED** [review]
Looks great!

**Review by raymond-hughes — APPROVED** [review]

**Pull request merged**